src/mono/wasm/runtime/assets.ts

```diff
@@ -181,52 +225,87 @@ async function start_asset_download_sources(asset: AssetEntry): Promise<AssetEnt
                 hash: asset.hash,
                 behavior: asset.behavior
             });
-            asset.pending = loadingResource;
+            asset.internalPending = loadingResource;
             response = await loadingResource.response;
             if (!response.ok) {
                 continue;// next source
             }
-            ++downloded_assets_count;
-            return asset;
+            ++actual_downloded_assets_count;
+            return response;
         }
         catch (err) {
             continue; //next source
         }
     }
-    throw response;
+    const isOkToFail = asset.isOptional || (asset.name.match(/\.pdb$/) && runtimeHelpers.config.ignorePdbLoadErrors);
+    mono_assert(response, "Response undefined");
+    if (!isOkToFail) {
+        const err: any = new Error(`MONO_WASM: download '${response.url}' for ${asset.name} failed ${response.status} ${response.statusText}`);
+        err.status = response.status;
+        throw err;
+    } else {
+        Module.print(`MONO_WASM: download '${response.url}' for ${asset.name} failed ${response.status} ${response.statusText}`);
+        if (!skipDownloadsAssetTypes[asset.behavior]) {
+            expected_downloded_assets_count--;
+        }
+        if (skipInstantiateAssetTypes[asset.behavior]) {
+            expected_instantiated_assets_count--;
+        }
+        return undefined;
+    }
 }
 
-async function start_asset_download_throttle(asset: AssetEntry): Promise<AssetEntry | undefined> {
-    // we don't addRunDependency to allow download in parallel with onRuntimeInitialized event!
-    while (throttling) {
-        await throttling.promise;
-    }
-    try {
-        ++parallel_count;
-        if (parallel_count == runtimeHelpers.maxParallelDownloads) {
-            if (runtimeHelpers.diagnosticTracing)
-                console.debug("MONO_WASM: Throttling further parallel downloads");
-            throttling = createPromiseController<void>();
+function resolve_path(asset: AssetEntry, sourcePrefix: string): string {
+    let attemptUrl;
+    const assemblyRootFolder = runtimeHelpers.config.assemblyRootFolder;
+    if (!asset.resolvedUrl) {
+        if (sourcePrefix === "") {
```

Comment:
this can't be undefined, right?

Reply:
it can't. added assert for it, just to make it clear.